File: packages/memo_stack_server/memo_stack_server/extraction_capabilities.py
"""Safe extraction capability payloads for public diagnostics."""

from __future__ import annotations

import importlib.util
from dataclasses import dataclass

from memo_stack_server.config import Settings

_PROFILE_ORDER = (
    "standard_local",
    "standard_docling",
    "standard_vision",
    "media_api",
    "media_local_asr",
    "standard_asr",
    "standard_full",
)

_PROFILE_CONTRACTS: dict[str, dict[str, tuple[str, ...]]] = {
    "standard_local": {
        "input_modalities": (
            "text",
            "document",
            "image",
            "timed_text",
            "audio_metadata",
            "video_metadata",
        ),
        "evidence_coordinates": (
            "char_range",
            "page_number",
            "bbox",
            "time_range_ms",
        ),
        "primary_artifact_types": (
            "image_regions",
            "transcript",
            "media_manifest",
            "keyframe",
            "video_frame_timeline",
        ),
        "document_features": ("plain_text", "pdf_text", "basic_metadata"),
        "vision_features": ("image_metadata", "ocr_regions_when_available"),
        "transcript_features": ("timed_text_segments", "time_ranges"),
        "video_features": ("ffprobe_metadata", "sampled_keyframes", "frame_timeline"),
    },
    "standard_docling": {
        "input_modalities": ("document",),
        "evidence_coordinates": ("char_range", "page_number", "bbox"),
        "primary_artifact_types": ("normalized_json", "table_html"),
        "document_features": (
            "layout",
            "reading_order",
            "tables",
            "ocr_when_enabled",
            "normalized_json",
        ),
    },
    "standard_vision": {
        "input_modalities": ("image",),
        "evidence_coordinates": ("bbox",),
        "primary_artifact_types": ("vision_json", "image_regions"),
        "vision_features": (
            "structured_image_summary",
            "detected_text",
            "region_coordinates",
            "provider_payload_bounding",
        ),
    },
    "media_api": {
        "input_modalities": ("audio", "video"),
        "evidence_coordinates": ("time_range_ms", "bbox"),
        "primary_artifact_types": (
            "transcript",
            "transcript_json",
            "keyframe",
            "video_frame_timeline",
        ),
        "transcript_features": (
            "segments",
            "time_ranges",
            "transcript_json",
            "optional_speaker_labels",
            "optional_word_timestamps",
        ),
        "video_features": ("ffprobe_metadata", "sampled_keyframes", "frame_timeline"),
    },
    "media_local_asr": {
        "input_modalities": ("audio", "video"),
        "evidence_coordinates": ("time_range_ms",),
        "primary_artifact_types": (
            "media_manifest",
            "transcript",
            "transcript_json",
        ),
        "transcript_features": ("segments", "time_ranges", "transcript_json"),
        "video_features": ("ffprobe_metadata",),
    },
    "standard_asr": {
        "input_modalities": ("audio", "video"),
        "evidence_coordinates": ("time_range_ms", "bbox"),
        "primary_artifact_types": (
            "transcript",
            "transcript_json",
            "keyframe",
            "video_frame_timeline",
        ),
        "transcript_features": (
            "segments",
            "time_ranges",
            "transcript_json",
            "optional_speaker_labels",
            "optional_word_timestamps",
        ),
        "video_features": ("ffprobe_metadata", "sampled_keyframes", "frame_timeline"),
    },
    "standard_full": {
        "input_modalities": ("document", "image", "audio", "video"),
        "evidence_coordinates": (
            "char_range",
            "page_number",
            "bbox",
            "time_range_ms",
        ),
        "primary_artifact_types": (
            "normalized_json",
            "table_html",
            "vision_json",
            "image_regions",
            "transcript",
            "transcript_json",
            "keyframe",
            "video_frame_timeline",
        ),
        "document_features": (
            "layout",
            "reading_order",
            "tables",
            "ocr_when_enabled",
            "normalized_json",
        ),
        "vision_features": (
            "structured_image_summary",
            "detected_text",
            "region_coordinates",
            "provider_payload_bounding",
        ),
        "transcript_features": (
            "segments",
            "time_ranges",
            "transcript_json",
            "optional_speaker_labels",
            "optional_word_timestamps",
        ),
        "video_features": ("ffprobe_metadata", "sampled_keyframes", "frame_timeline"),
    },
}


@dataclass(frozen=True)
class _ProviderState:
    name: str
    kind: str
    installed: bool
    configured: bool
    enabled: bool
    status: str
    reason: str | None
    profiles: tuple[str, ...]
    external_provider_egress: bool
    metadata: dict[str, object]

    def as_public_dict(self) -> dict[str, object]:
        payload: dict[str, object] = {
            "kind": self.kind,
            "installed": self.installed,
            "configured": self.configured,
            "enabled": self.enabled,
            "status": self.status,
            "profiles": list(self.profiles),
            "external_provider_egress": self.external_provider_egress,
        }
        if self.reason is not None:
            payload["reason"] = self.reason
        payload.update(self.metadata)
        return payload


def build_extraction_capability_payload(settings: Settings) -> dict[str, object]:
    providers = _provider_states(settings)
    profiles = _profile_states(settings, providers)
    return {
        "enabled": settings.extraction_enabled,
        "default_profile": settings.extraction_default_profile,
        "profiles": list(_PROFILE_ORDER),
        "profiles_v2": [profiles[name] for name in _PROFILE_ORDER],
        "providers": {
            name: provider.as_public_dict()
            for name, provider in sorted(providers.items(), key=lambda item: item[0])
        },
        "optional_extras": _legacy_optional_extras(settings, providers),
        "policy": _policy_payload(settings),
        "evidence_contract": _evidence_contract_payload(),
        "feature_contract": _feature_contract_payload(),
        "external_provider_egress": settings.extraction_external_ai_enabled,
        "limits": _limits_payload(settings),
    }


def _provider_states(settings: Settings) -> dict[str, _ProviderState]:
    external_ready = settings.extraction_external_ai_enabled and bool(settings.openai_api_key)
    openai_installed = _module_available("openai")
    docling_installed = _module_available("docling")
    local_asr_installed = _module_available("faster_whisper")
    transcription_configured = (
        settings.transcription_provider == "openai" and external_ready and openai_installed
    )
    vision_configured = external_ready and openai_installed
    return {
        "docling": _ProviderState(
            name="docling",
            kind="document_parser",
            installed=docling_installed,
            configured=docling_installed,
            enabled=settings.extraction_enabled and docling_installed,
            status=_status(settings.extraction_enabled, docling_installed, docling_installed),
            reason=None if docling_installed else "provider_package_missing",
            profiles=("standard_docling", "standard_full"),
            external_provider_egress=False,
            metadata={},
        ),
        "openai_vision": _ProviderState(
            name="openai_vision",
            kind="image_vision",
            installed=openai_installed,
            configured=vision_configured,
            enabled=settings.extraction_enabled and vision_configured,
            status=_status(settings.extraction_enabled, openai_installed, vision_configured),
            reason=_external_reason(
                installed=openai_installed,
                external_ai_enabled=settings.extraction_external_ai_enabled,
                credential_present=bool(settings.openai_api_key),
            ),
            profiles=("standard_vision", "standard_full"),
            external_provider_egress=True,
            metadata={
                "model": settings.extraction_vision_model,
                "detail": settings.extraction_vision_detail,
            },
        ),
        "transcription_api": _ProviderState(
            name="transcription_api",
            kind="speech_to_text",
            installed=openai_installed,
            configured=transcription_configured,
            enabled=settings.extraction_enabled and transcription_configured,
            status=_status(settings.extraction_enabled, openai_installed, transcription_configured),
            reason=_transcription_reason(
                installed=openai_installed,
                provider=settings.transcription_provider,
                external_ai_enabled=settings.extraction_external_ai_enabled,
                credential_present=bool(settings.openai_api_key),
            ),
            profiles=("media_api", "standard_asr", "standard_full"),
            external_provider_egress=True,
            metadata={
                "provider": settings.transcription_provider,
                "model": settings.transcription_openai_model,
                "max_provider_upload_bytes": settings.transcription_openai_max_upload_bytes,
                "diarization_model_configured": _transcription_model_supports_diarization(
                    settings.transcription_openai_model
                ),
            },
        ),
        "transcription_local": _ProviderState(
            name="transcription_local",
            kind="speech_to_text",
            installed=local_asr_installed,
            configured=local_asr_installed,
            enabled=settings.extraction_enabled and local_asr_installed,
            status=_status(settings.extraction_enabled, local_asr_installed, local_asr_installed),
            reason=None if local_asr_installed else "provider_package_missing",
            profiles=("media_local_asr", "asr:<model>", "faster_whisper:<model>"),
            external_provider_egress=False,
            metadata={
                "model": settings.extraction_asr_model,
                "device": settings.extraction_asr_device,
                "compute_type": settings.extraction_asr_compute_type,
                "default": False,
            },
        ),
    }


def _profile_states(
    settings: Settings,
    providers: dict[str, _ProviderState],
) -> dict[str, dict[str, object]]:
    extraction_enabled = settings.extraction_enabled
    return {
        "standard_local": _profile_payload(
            name="standard_local",
            enabled=extraction_enabled,
            status="ok" if extraction_enabled else "disabled",
            reason=None if extraction_enabled else "extraction_disabled",
            provider_names=("local_text", "pdf_text", "image_metadata", "media_metadata"),
            external_provider_egress=False,
            requires_explicit_external_ai=False,
            fallback_profiles=(),
        ),
        "standard_docling": _profile_from_providers(
            name="standard_docling",
            providers=(providers["docling"],),
            fallback_profiles=("standard_local",),
            requires_explicit_external_ai=False,
        ),
        "standard_vision": _profile_from_providers(
            name="standard_vision",
            providers=(providers["openai_vision"],),
            fallback_profiles=("standard_local",),
            requires_explicit_external_ai=True,
        ),
        "media_api": _profile_from_providers(
            name="media_api",
            providers=(providers["transcription_api"],),
            fallback_profiles=("standard_local",),
            requires_explicit_external_ai=True,
        ),
        "media_local_asr": _profile_from_providers(
            name="media_local_asr",
            providers=(providers["transcription_local"],),
            fallback_profiles=("standard_local",),
            requires_explicit_external_ai=False,
            may_run_local_asr=True,
        ),
        "standard_asr": _profile_from_providers(
            name="standard_asr",
            providers=(providers["transcription_api"],),
            fallback_profiles=("standard_local",),
            requires_explicit_external_ai=True,
            deprecated=True,
            replacement_profiles=("media_api", "media_local_asr"),
        ),
        "standard_full": _profile_from_providers(
            name="standard_full",
            providers=(
                providers["docling"],
                providers["openai_vision"],
                providers["transcription_api"],
            ),
            fallback_profiles=(
                "standard_docling",
                "standard_vision",
                "media_api",
                "standard_local",
            ),
            requires_explicit_external_ai=True,
        ),
    }


def _profile_from_providers(
    *,
    name: str,
    providers: tuple[_ProviderState, ...],
    fallback_profiles: tuple[str, ...],
    requires_explicit_external_ai: bool,
    deprecated: bool = False,
    may_run_local_asr: bool = False,
    replacement_profiles: tuple[str, ...] = (),
) -> dict[str, object]:
    enabled_providers = tuple(provider for provider in providers if provider.enabled)
    status = _combined_status(providers)
    reason = _combined_reason(providers, status)
    return _profile_payload(
        name=name,
        enabled=bool(enabled_providers),
        status=status,
        reason=reason,
        provider_names=tuple(provider.name for provider in providers),
        external_provider_egress=any(provider.external_provider_egress for provider in providers),
        requires_explicit_external_ai=requires_explicit_external_ai,
        fallback_profiles=fallback_profiles,
        deprecated=deprecated,
        may_run_local_asr=may_run_local_asr,
        replacement_profiles=replacement_profiles,
    )


def _profile_payload(
    *,
    name: str,
    enabled: bool,
    status: str,
    reason: str | None,
    provider_names: tuple[str, ...],
    external_provider_egress: bool,
    requires_explicit_external_ai: bool,
    fallback_profiles: tuple[str, ...],
    deprecated: bool = False,
    may_run_local_asr: bool = False,
    replacement_profiles: tuple[str, ...] = (),
) -> dict[str, object]:
    contract = _PROFILE_CONTRACTS.get(
        name,
        {
            "input_modalities": (),
            "evidence_coordinates": (),
            "primary_artifact_types": (),
            "document_features": (),
            "vision_features": (),
            "transcript_features": (),
            "video_features": (),
        },
    )
    payload: dict[str, object] = {
        "name": name,
        "enabled": enabled,
        "status": status,
        "providers": list(provider_names),
        "input_modalities": list(contract["input_modalities"]),
        "evidence_coordinates": list(contract["evidence_coordinates"]),
        "primary_artifact_types": list(contract["primary_artifact_types"]),
        "document_features": list(contract.get("document_features", ())),
        "vision_features": list(contract.get("vision_features", ())),
        "transcript_features": list(contract.get("transcript_features", ())),
        "video_features": list(contract.get("video_features", ())),
        "external_provider_egress": external_provider_egress,
        "requires_explicit_external_ai": requires_explicit_external_ai,
        "fallback_profiles": list(fallback_profiles),
        "memory_promotion": "review_required",
        "source_text_policy": "untrusted_evidence",
        "artifact_payloads_bounded": True,
        "may_run_local_asr": may_run_local_asr,
    }
    if reason is not None:
        payload["reason"] = reason
    if deprecated:
        payload["deprecated"] = True
    if replacement_profiles:
        payload["replacement_profiles"] = list(replacement_profiles)
    return payload


def _legacy_optional_extras(
    settings: Settings,
    providers: dict[str, _ProviderState],
) -> dict[str, dict[str, object]]:
    return {
        "docling": {
            "installed": providers["docling"].installed,
            "profiles": ["standard_docling", "standard_full"],
        },
        "vision": {
            "installed": providers["openai_vision"].installed,
            "configured": providers["openai_vision"].configured,
            "profiles": ["standard_vision", "standard_full"],
            "model": settings.extraction_vision_model,
            "detail": settings.extraction_vision_detail,
        },
        "transcription_api": {
            "installed": providers["transcription_api"].installed,
            "configured": providers["transcription_api"].configured,
            "provider": settings.transcription_provider,
            "profiles": ["media_api", "standard_asr", "standard_full"],
            "model": settings.transcription_openai_model,
            "max_provider_upload_bytes": settings.transcription_openai_max_upload_bytes,
            "diarization_model_configured": _transcription_model_supports_diarization(
                settings.transcription_openai_model
            ),
        },
        "transcription_local": {
            "installed": providers["transcription_local"].installed,
            "profiles": ["media_local_asr", "asr:<model>", "faster_whisper:<model>"],
            "model": settings.extraction_asr_model,
            "device": settings.extraction_asr_device,
            "compute_type": settings.extraction_asr_compute_type,
            "default": False,
        },
        "asr": {
            "deprecated": True,
            "replacement_profiles": ["media_api", "media_local_asr"],
        },
    }


def _policy_payload(settings: Settings) -> dict[str, object]:
    return {
        "schema_version": 2,
        "external_ai_allowed": settings.extraction_external_ai_enabled,
        "external_ai_requires_explicit_profile": True,
        "local_asr_default": False,
        "local_asr_requires_explicit_profile": True,
        "memory_promotion": "review_required",
        "source_text_policy": "untrusted_evidence",
        "provider_payloads_bounded": True,
        "sensitive_data_in_diagnostics": False,
        "canonical_store": "postgres",
        "derived_indexes": ["qdrant", "graphiti"],
    }


def _evidence_contract_payload() -> dict[str, object]:
    return {
        "schema_version": "memo_stack.extraction_evidence_contract.v1",
        "source_ref_coordinate_fields": [
            "char_start",
            "char_end",
            "page_number",
            "bbox",
            "time_start_ms",
            "time_end_ms",
        ],
        "profile_contract_fields": [
            "input_modalities",
            "evidence_coordinates",
            "primary_artifact_types",
        ],
        "coordinates_are_optional_per_item": True,
        "source_refs_are_bounded": True,
        "memory_promotion": "review_required",
        "source_text_policy": "untrusted_evidence",
    }


def _feature_contract_payload() -> dict[str, object]:
    return {
        "schema_version": "memo_stack.extraction_feature_contract.v1",
        "profile_feature_fields": [
            "document_features",
            "vision_features",
            "transcript_features",
            "video_features",
        ],
        "feature_values_are_capabilities_not_guarantees": True,
        "actual_artifact_metadata_is_authoritative": True,
        "external_ai_features_require_explicit_profile": True,
        "local_asr_does_not_provide_speaker_labels": True,
    }


def _limits_payload(settings: Settings) -> dict[str, object]:
    return {
        "max_bytes": settings.extraction_max_bytes,
        "max_pages": settings.extraction_max_pages,
        "max_media_seconds": settings.extraction_max_media_seconds,
        "max_output_chars": settings.extraction_max_output_chars,
        "max_tables": settings.extraction_max_tables,
        "ocr_enabled": settings.extraction_ocr_enabled,
        "max_image_pixels": settings.extraction_max_image_pixels,
        "parser_timeout_seconds": settings.extraction_parser_timeout_seconds,
        "subprocess_timeout_seconds": settings.extraction_subprocess_timeout_seconds,
    }


def _status(extraction_enabled: bool, installed: bool, configured: bool) -> str:
    if not extraction_enabled:
        return "disabled"
    if not installed:
        return "unavailable"
    if not configured:
        return "blocked"
    return "ok"


def _combined_status(providers: tuple[_ProviderState, ...]) -> str:
    statuses = {provider.status for provider in providers}
    if "ok" in statuses:
        return "ok" if statuses == {"ok"} else "degraded"
    if "blocked" in statuses:
        return "blocked"
    if "unavailable" in statuses:
        return "unavailable"
    return "disabled"


def _combined_reason(providers: tuple[_ProviderState, ...], status: str) -> str | None:
    if status == "ok":
        return None
    reasons = tuple(provider.reason for provider in providers if provider.reason)
    if not reasons:
        return status
    return reasons[0]


def _external_reason(
    *,
    installed: bool,
    external_ai_enabled: bool,
    credential_present: bool,
) -> str | None:
    if not installed:
        return "provider_package_missing"
    if not external_ai_enabled:
        return "external_ai_disabled"
    if not credential_present:
        return "provider_credential_missing"
    return None


def _transcription_reason(
    *,
    installed: bool,
    provider: str,
    external_ai_enabled: bool,
    credential_present: bool,
) -> str | None:
    if provider == "disabled":
        return "provider_disabled"
    return _external_reason(
        installed=installed,
        external_ai_enabled=external_ai_enabled,
        credential_present=credential_present,
    )


def _transcription_model_supports_diarization(model: str) -> bool:
    return "diarize" in model.strip().lower()


def _module_available(module_name: str) -> bool:
    return importlib.util.find_spec(module_name) is not None

File: frontend/lib/src/features/chat/domain/entities/extraction_capabilities.dart
import 'package:equatable/equatable.dart';

class ExtractionCapabilities extends Equatable {
  final bool enabled;
  final String? defaultProfile;
  final Map<String, ExtractionProfileCapability> profiles;
  final Map<String, ExtractionProviderCapability> providers;
  final Map<String, Map<String, ExtractionModalityAction>> modalityActions;
  final List<ExtractionDegradedComponent> degradedComponents;
  final Map<String, dynamic> policy;
  final Map<String, dynamic> evidenceContract;
  final Map<String, dynamic> featureContract;
  final Map<String, dynamic> providerContract;
  final Map<String, dynamic> manifestContract;
  final Map<String, dynamic> fileTypeDetection;
  final Map<String, dynamic> limits;
  final Map<String, dynamic> raw;

  const ExtractionCapabilities({
    required this.enabled,
    required this.defaultProfile,
    required this.profiles,
    required this.providers,
    required this.modalityActions,
    required this.degradedComponents,
    required this.policy,
    required this.evidenceContract,
    required this.featureContract,
    required this.providerContract,
    required this.manifestContract,
    required this.fileTypeDetection,
    required this.limits,
    required this.raw,
  });

  factory ExtractionCapabilities.fromMap(Map<String, dynamic> map) {
    final profiles = {
      for (final item in _listOfMaps(map['profiles_v2']))
        if (_string(item['name']).isNotEmpty)
          _string(item['name']): ExtractionProfileCapability.fromMap(item),
    };
    return ExtractionCapabilities(
      enabled: map['enabled'] == true,
      defaultProfile: _nullableString(map['default_profile']),
      profiles: profiles,
      providers: _providerMap(map['providers']),
      modalityActions: _modalityActionMap(map['modality_actions']),
      degradedComponents: _listOfMaps(map['degraded_components'])
          .map(ExtractionDegradedComponent.fromMap)
          .toList(growable: false),
      policy: _map(map['policy']),
      evidenceContract: _map(map['evidence_contract']),
      featureContract: _map(map['feature_contract']),
      providerContract: _map(map['provider_contract']),
      manifestContract: _map(map['manifest_contract']),
      fileTypeDetection: _map(map['file_type_detection']),
      limits: _map(map['limits']),
      raw: Map<String, dynamic>.from(map),
    );
  }

  ExtractionProfileCapability? profile(String name) => profiles[name];
  ExtractionProviderCapability? provider(String name) => providers[name];

  ExtractionModalityAction? modalityAction(String modality, String action) {
    return modalityActions[modality]?[action];
  }

  bool get hasDegradedComponents => degradedComponents.isNotEmpty;

  List<String> get degradedLabels => degradedComponents
      .map((item) => item.label)
      .where((item) => item.isNotEmpty)
      .toList(growable: false);

  @override
  List<Object?> get props => [
        enabled,
        defaultProfile,
        profiles,
        providers,
        modalityActions,
        degradedComponents,
        policy,
        evidenceContract,
        featureContract,
        providerContract,
        manifestContract,
        fileTypeDetection,
        limits,
        raw,
      ];
}

class ExtractionProfileCapability extends Equatable {
  final String name;
  final bool enabled;
  final String status;
  final String? reason;
  final List<String> providers;
  final List<String> inputModalities;
  final List<String> evidenceCoordinates;
  final List<String> primaryArtifactTypes;
  final List<String> documentFeatures;
  final List<String> visionFeatures;
  final List<String> transcriptFeatures;
  final List<String> videoFeatures;
  final bool externalProviderEgress;
  final bool requiresExplicitExternalAi;
  final bool mayRunLocalAsr;
  final bool deprecated;
  final List<String> fallbackProfiles;
  final List<String> replacementProfiles;
  final String memoryPromotion;
  final String sourceTextPolicy;
  final bool artifactPayloadsBounded;
  final Map<String, dynamic> raw;

  const ExtractionProfileCapability({
    required this.name,
    required this.enabled,
    required this.status,
    required this.reason,
    required this.providers,
    required this.inputModalities,
    required this.evidenceCoordinates,
    required this.primaryArtifactTypes,
    required this.documentFeatures,
    required this.visionFeatures,
    required this.transcriptFeatures,
    required this.videoFeatures,
    required this.externalProviderEgress,
    required this.requiresExplicitExternalAi,
    required this.mayRunLocalAsr,
    required this.deprecated,
    required this.fallbackProfiles,
    required this.replacementProfiles,
    required this.memoryPromotion,
    required this.sourceTextPolicy,
    required this.artifactPayloadsBounded,
    required this.raw,
  });

  factory ExtractionProfileCapability.fromMap(Map<String, dynamic> map) {
    return ExtractionProfileCapability(
      name: _string(map['name']),
      enabled: map['enabled'] == true,
      status: _string(map['status'], fallback: 'unknown'),
      reason: _nullableString(map['reason']),
      providers: _stringList(map['providers']),
      inputModalities: _stringList(map['input_modalities']),
      evidenceCoordinates: _stringList(map['evidence_coordinates']),
      primaryArtifactTypes: _stringList(map['primary_artifact_types']),
      documentFeatures: _stringList(map['document_features']),
      visionFeatures: _stringList(map['vision_features']),
      transcriptFeatures: _stringList(map['transcript_features']),
      videoFeatures: _stringList(map['video_features']),
      externalProviderEgress: map['external_provider_egress'] == true,
      requiresExplicitExternalAi: map['requires_explicit_external_ai'] == true,
      mayRunLocalAsr: map['may_run_local_asr'] == true,
      deprecated: map['deprecated'] == true,
      fallbackProfiles: _stringList(map['fallback_profiles']),
      replacementProfiles: _stringList(map['replacement_profiles']),
      memoryPromotion: _string(map['memory_promotion']),
      sourceTextPolicy: _string(map['source_text_policy']),
      artifactPayloadsBounded: map['artifact_payloads_bounded'] == true,
      raw: Map<String, dynamic>.from(map),
    );
  }

  @override
  List<Object?> get props => [
        name,
        enabled,
        status,
        reason,
        providers,
        inputModalities,
        evidenceCoordinates,
        primaryArtifactTypes,
        documentFeatures,
        visionFeatures,
        transcriptFeatures,
        videoFeatures,
        externalProviderEgress,
        requiresExplicitExternalAi,
        mayRunLocalAsr,
        deprecated,
        fallbackProfiles,
        replacementProfiles,
        memoryPromotion,
        sourceTextPolicy,
        artifactPayloadsBounded,
        raw,
      ];
}

class ExtractionProviderCapability extends Equatable {
  final String name;
  final String kind;
  final bool installed;
  final bool configured;
  final bool enabled;
  final String status;
  final String? reason;
  final List<String> profiles;
  final bool externalProviderEgress;
  final String? operatorAction;
  final bool? userRetryable;
  final Map<String, dynamic> raw;

  const ExtractionProviderCapability({
    required this.name,
    required this.kind,
    required this.installed,
    required this.configured,
    required this.enabled,
    required this.status,
    required this.reason,
    required this.profiles,
    required this.externalProviderEgress,
    required this.operatorAction,
    required this.userRetryable,
    required this.raw,
  });

  factory ExtractionProviderCapability.fromMap(
    String name,
    Map<String, dynamic> map,
  ) {
    return ExtractionProviderCapability(
      name: name,
      kind: _string(map['kind']),
      installed: map['installed'] == true,
      configured: map['configured'] == true,
      enabled: map['enabled'] == true,
      status: _string(map['status'], fallback: 'unknown'),
      reason: _nullableString(map['reason']),
      profiles: _stringList(map['profiles']),
      externalProviderEgress: map['external_provider_egress'] == true,
      operatorAction: _nullableString(map['operator_action']),
      userRetryable: _nullableBool(map['user_retryable']),
      raw: Map<String, dynamic>.from(map),
    );
  }

  bool get isDegraded =>
      status == 'blocked' || status == 'unavailable' || status == 'degraded';

  String get label {
    final action = operatorAction;
    if (action != null && action.isNotEmpty) return '$name: $action';
    final text = reason;
    if (text != null && text.isNotEmpty) return '$name: $text';
    return '$name: $status';
  }

  @override
  List<Object?> get props => [
        name,
        kind,
        installed,
        configured,
        enabled,
        status,
        reason,
        profiles,
        externalProviderEgress,
        operatorAction,
        userRetryable,
        raw,
      ];
}

class ExtractionModalityAction extends Equatable {
  final String modality;
  final String action;
  final bool enabled;
  final String status;
  final String? reason;
  final List<String> profiles;
  final List<String> providers;
  final List<String> artifactTypes;
  final List<String> evidenceCoordinates;
  final bool externalProviderEgress;
  final bool requiresExplicitExternalAi;
  final List<String> fallbackProfiles;
  final String memoryPromotion;
  final String sourceTextPolicy;
  final bool artifactPayloadsBounded;
  final String? operatorAction;
  final bool? userRetryable;
  final Map<String, dynamic> raw;

  const ExtractionModalityAction({
    required this.modality,
    required this.action,
    required this.enabled,
    required this.status,
    required this.reason,
    required this.profiles,
    required this.providers,
    required this.artifactTypes,
    required this.evidenceCoordinates,
    required this.externalProviderEgress,
    required this.requiresExplicitExternalAi,
    required this.fallbackProfiles,
    required this.memoryPromotion,
    required this.sourceTextPolicy,
    required this.artifactPayloadsBounded,
    required this.operatorAction,
    required this.userRetryable,
    required this.raw,
  });

  factory ExtractionModalityAction.fromMap(
    String modality,
    String action,
    Map<String, dynamic> map,
  ) {
    return ExtractionModalityAction(
      modality: modality,
      action: action,
      enabled: map['enabled'] == true,
      status: _string(map['status'], fallback: 'unknown'),
      reason: _nullableString(map['reason']),
      profiles: _stringList(map['profiles']),
      providers: _stringList(map['providers']),
      artifactTypes: _stringList(map['artifact_types']),
      evidenceCoordinates: _stringList(map['evidence_coordinates']),
      externalProviderEgress: map['external_provider_egress'] == true,
      requiresExplicitExternalAi: map['requires_explicit_external_ai'] == true,
      fallbackProfiles: _stringList(map['fallback_profiles']),
      memoryPromotion: _string(map['memory_promotion']),
      sourceTextPolicy: _string(map['source_text_policy']),
      artifactPayloadsBounded: map['artifact_payloads_bounded'] != false,
      operatorAction: _nullableString(map['operator_action']),
      userRetryable: _nullableBool(map['user_retryable']),
      raw: Map<String, dynamic>.from(map),
    );
  }

  bool get isDegraded =>
      !enabled ||
      status == 'blocked' ||
      status == 'unavailable' ||
      status == 'degraded' ||
      status == 'disabled';

  String get label {
    final actionText = operatorAction;
    if (actionText != null && actionText.isNotEmpty) {
      return '$modality.$action: $actionText';
    }
    final reasonText = reason;
    if (reasonText != null && reasonText.isNotEmpty) {
      return '$modality.$action: $reasonText';
    }
    return '$modality.$action: $status';
  }

  @override
  List<Object?> get props => [
        modality,
        action,
        enabled,
        status,
        reason,
        profiles,
        providers,
        artifactTypes,
        evidenceCoordinates,
        externalProviderEgress,
        requiresExplicitExternalAi,
        fallbackProfiles,
        memoryPromotion,
        sourceTextPolicy,
        artifactPayloadsBounded,
        operatorAction,
        userRetryable,
        raw,
      ];
}

class ExtractionDegradedComponent extends Equatable {
  final String componentType;
  final String name;
  final String status;
  final String? reason;
  final String? operatorAction;
  final bool? userRetryable;
  final Map<String, dynamic> raw;

  const ExtractionDegradedComponent({
    required this.componentType,
    required this.name,
    required this.status,
    required this.reason,
    required this.operatorAction,
    required this.userRetryable,
    required this.raw,
  });

  factory ExtractionDegradedComponent.fromMap(Map<String, dynamic> map) {
    return ExtractionDegradedComponent(
      componentType: _string(map['component_type']),
      name: _string(map['name']),
      status: _string(map['status'], fallback: 'unknown'),
      reason: _nullableString(map['reason']),
      operatorAction: _nullableString(map['operator_action']),
      userRetryable: _nullableBool(map['user_retryable']),
      raw: Map<String, dynamic>.from(map),
    );
  }

  String get label {
    final target = name.isEmpty ? componentType : name;
    final action = operatorAction;
    if (action != null && action.isNotEmpty) return '$target: $action';
    final text = reason;
    if (text != null && text.isNotEmpty) return '$target: $text';
    return '$target: $status';
  }

  @override
  List<Object?> get props => [
        componentType,
        name,
        status,
        reason,
        operatorAction,
        userRetryable,
        raw,
      ];
}

Map<String, ExtractionProviderCapability> _providerMap(Object? value) {
  final raw = _map(value);
  return {
    for (final entry in raw.entries)
      if (entry.value is Map)
        entry.key: ExtractionProviderCapability.fromMap(
          entry.key,
          _map(entry.value),
        ),
  };
}

Map<String, Map<String, ExtractionModalityAction>> _modalityActionMap(
  Object? value,
) {
  final raw = _map(value);
  final parsed = <String, Map<String, ExtractionModalityAction>>{};
  for (final modalityEntry in raw.entries) {
    final actions = _map(modalityEntry.value);
    final parsedActions = <String, ExtractionModalityAction>{};
    for (final actionEntry in actions.entries) {
      if (actionEntry.value is! Map) continue;
      parsedActions[actionEntry.key] = ExtractionModalityAction.fromMap(
        modalityEntry.key,
        actionEntry.key,
        _map(actionEntry.value),
      );
    }
    if (parsedActions.isNotEmpty) parsed[modalityEntry.key] = parsedActions;
  }
  return parsed;
}

Map<String, dynamic> _map(Object? value) {
  if (value is Map<String, dynamic>) return Map<String, dynamic>.from(value);
  if (value is Map) {
    return value.map((key, item) => MapEntry(key.toString(), item));
  }
  return const <String, dynamic>{};
}

List<Map<String, dynamic>> _listOfMaps(Object? value) {
  if (value is! List) return const <Map<String, dynamic>>[];
  return value
      .whereType<Map>()
      .map((item) => item.map((key, value) => MapEntry(key.toString(), value)))
      .toList(growable: false);
}

List<String> _stringList(Object? value) {
  if (value is! List) return const <String>[];
  return value
      .map((item) => item.toString().trim())
      .where((item) => item.isNotEmpty)
      .toList(growable: false);
}

String _string(Object? value, {String fallback = ''}) {
  final text = value?.toString().trim();
  if (text == null || text.isEmpty) return fallback;
  return text;
}

String? _nullableString(Object? value) {
  final text = value?.toString().trim();
  if (text == null || text.isEmpty) return null;
  return text;
}

bool? _nullableBool(Object? value) {
  if (value is bool) return value;
  return null;
}
